### PR TITLE
runtime: Add multi-block end-to-end test for SIMD-0123 edge cases

### PR DESCRIPTION
**Pull request by joncinque**
#### Problem

SIMD-0123 is implemented, but most of the tests are unit tests, with nothing testing all of the edge cases in one place.

#### Summary of changes

Add that test! This follows the inflation reward tests, but focuses on all the important edge cases to consider for SIMD-0123.

This PR also comprises two commits, the first of which refactors the bank setup and adds an activating and deactivating case. Those can eventually be split up into separate PRs if desired.

#### Testing

The new test.

~~Note: This builds on #14946, so leaving in draft until that lands. If reviewing, the two last commits are this PR. The final commit contains the new test.~~

### CODE REVIEW

**Comment by joncinque** [comment]
This is ready for a real review now

**Review by buffalojoec — APPROVED** [review]
Nice, gud test.

File: runtime/src/bank/partitioned_epoch_rewards/mod.rs [hunk 1]
```diff
@@ -701,6 +708,307 @@ mod tests {
...
+
+        // Deactivating stake functions like activated stake
+        let deactivating = add_vote_and_stake_accounts_with_activation_and_deactivation_epochs(
+            &bank,
+            &mut voters,
+            &mut stakers,
+            stake_lamports,
+            vote_lamports + pending_delegator_rewards,
+            0,
+            bank.epoch(),
```
Comment (buffalojoec) [inline runtime/src/bank/partitioned_epoch_rewards/mod.rs:972]: Are these both 0 for the new test? Are you sure this actually does a deactivating stake account?

**Review by buffalojoec — COMMENTED** [review]
Sorry @joncinque for some reason this one comment didn't send.

File: runtime/src/bank/partitioned_epoch_rewards/mod.rs (hunk 1 above)
Comment (joncinque) [reply runtime/src/bank/partitioned_epoch_rewards/mod.rs:972]: Oh yeah good call, this would just be never active. Changed to `Epoch::MAX` so that it's a bootstrap stake in #15430
